### PR TITLE
2.0: rewrite cartesian/ and numberAxis/ by ts

### PR DESCRIPTION
**Pull request by xile611**

### CODE REVIEW

**Comment by sainthkh** [comment]
Build failed.

**Comment by sainthkh** [comment]
I'll review them tomorrow.

**Comment by xile611** [comment]
@sainthkh 
I think It's better to keep the default behavior of `filterProps` consistent with `getPresentationAttributes` which will make it easier to migrate from `getPresentationAttributes` to `filterProps`. 

And add a optional parameter `includeEvents` to `filterProps` to allow filter all the presentation attributes and events.

This is the main change in the latest commit.

**Comment by sainthkh** [comment]
@xile611 Then, we need to make a plan for removing `PRESENTATION_ATTRIBUTES`.

File: src/cartesian/Area.tsx [hunk 1]
```diff
@@ -1,79 +1,90 @@
...
+interface AreaPointItem extends CurvePoint {
+  value?: number | number[];
+  payload?: any;
+}
 
 
-class Area extends PureComponent {
+interface InternalAreaProps {
+  xAxis?: Omit<XAxisProps, 'scale'> & { scale: D3Scale<string | number> };
+  yAxis?: Omit<YAxisProps, 'scale'> & { scale: D3Scale<string | number> };
```
Comment (sainthkh) [inline src/cartesian/Area.tsx]: I found this pattern a lot in your code. How about creating `InternalXAxisProps` and `InternalYAxisProps`?

File: src/cartesian/Area.tsx [hunk 2]
```diff
@@ -95,10 +106,10 @@ class Area extends PureComponent {
     animationEasing: 'ease',
   };
 
-  static getBaseValue = (props, xAxis, yAxis) => {
+  static getBaseValue = (props: Props, xAxis: Props['xAxis'], yAxis: Props['yAxis']): number => {
```
Comment (sainthkh) [inline src/cartesian/Area.tsx]: If you agree to create `InternalXAxisProps` and `InternalYAxisProps`, we can change `Props[`xAxis`]` and `Props[`yAxis`]` to them.

File: src/cartesian/Area.tsx [hunk 3]
```diff
@@ -120,7 +131,20 @@ class Area extends PureComponent {
...
   static getComposedData = ({ props, xAxis, yAxis, xAxisTicks, yAxisTicks, bandSize,
-    dataKey, stackedData, dataStartIndex, displayedData, offset }) => {
+    dataKey, stackedData, dataStartIndex, displayedData, offset }: {
+      props: Props;
+      item: Area;
+      bandSize: number;
+      xAxis: InternalAreaProps['xAxis'];
+      yAxis: InternalAreaProps['yAxis'];
+      xAxisTicks: XAxisProps['ticks'];
+      yAxisTicks: YAxisProps['ticks'];
```
Comment (sainthkh) [inline src/cartesian/Area.tsx]: I also see `XAxisProps['ticks']` pattern a lot. What do you think is better, `XAxisProps['ticks']` or `TickItem[]`?

File: src/cartesian/Area.tsx (hunk 2 above)
Comment (xile611) [reply src/cartesian/Area.tsx]: @sainthkh If we add `InternalXAxisProps` in `utils/types.ts`, `utils/types.ts` need to import `XAxis`, but `XAxis` need to import `utils/types.ts`. It leads to a cycle.

Comment (sainthkh) [reply src/cartesian/Area.tsx]: What I was talking about is putting them under `/cartesian/types.ts`. Because those `props` are only used in `/cartesian`.

Comment (sainthkh) [reply src/cartesian/Area.tsx]: Good. Let's leave them as-is for now. We need to think about them later.

**Pull request merged**